doc/developer/arrangements.md

```diff
@@ -0,0 +1,129 @@
+
+Let's talk through what an arrangement is, which operators require arrangements, and what we can do about them.
+
+## What is an "Arrangement"?
+
+An arrangement is an alternative representation of a stream of update triples `(data, time, diff)`.
+In stream form, the triple pass by and are gone.
+An arrangement records these triples, and indexes them by `data`.
+With some caveats about "logical compaction", the index maintains the *current* accumulation of `diff` for each `(data, time)`.
+This means that the size of an arrangement should be roughly the current size of the accumulated updates, where "zero" accumulations contribute nothing.
+An arrangement undergoes "physical compaction" lazily, which means that the actual size of an arrangement can exceed the logical size, for what should be a bounded amount of time.
+
+An arrangement only costs proportional to the number of distinct `(data, time)` pairs.
+This number can be small even if the number of records is large, if we were able to reduce `data` to few values.
+For example, if we want to histogram taxi rides by number of occupants and fare amount, there may be few distinct pairs despite many records.
+
+Arrangements are actually indexed assuming that `data` has a `(key, val)` structure.
+The index is by `key`, which will be important in understanding when arrangements can be shared and reused.
+Materialize will attempt to choose a `key` to ensure that the data are well distributed.
+For example, if a collection has a primary key, Materialize will default to this key for an arrangement.
+Each of `key` and `val` can be `()`, corresponding to "no key" and "the key is the value", respectively.
+
+To investigate the existing arrangements, query the `mz_arrangement_sizes` logging source.
+There are several diagnostic views in `mz_catalog` that connect this information to operator names, and group it by dataflow.
+
+## Which operators need Arrangements?
+
+We create more arrangements that just those to house materialized sources and views.
+Many internal operators require arrangements so that they can respond efficiently to changes.
+For example, the differential dataflow `join` operator requires that each of its two inputs be an arrangement.
+Specifically, each arrangement must have as its `key` the fields that will be equated by the join.
+Additionally, the differential dataflow `reduce` operator requires that both its input and its output be an arrangement.
+These two operators are at the heart of the arrangement use in Materialize, because we construct relational operators out of them.
+
+Materialize has several relational operators that are "rendered" using differential `join` and `reduce` operators.
+These are the operators you would see in `explain plan for <query>`.
+There is a more complicated mapping from SQL to these operators.
+
+---
+
+Our `Join` operator joins together a sequence of input relations.
```

Comment:
You're implying that the cost of a join proportionate to the total size of arrangements being maintained for each set of intermediate results, but I think it'd be clearer to be explicit about the total memory cost for the join. This is one of the bigger differences between the cost model for Materialize vs. a RDBMS.